### DISCUSSION TITLE
Create more Project API keys?

### DISCUSSION THREAD

**Discussion by MariuzM** [supabase/supabase#16130] [Questions] [answered]
So we have `anon` and `service_role`. I want to use my own BE to access some tables in SB.

Issue i have is when i use `service_role` then it `overrides` the RLS for auth only users to see table, and offcourse i cant use `anon` key because its also being used in public.

I want to generate another key that will be used on BE but also to be `auth` level, meaning that only signed in users can access table when they call my BE api.

**Comment by GaryAustin1** [comment, marked as answer]
You can mint a jwt with `authenticated` role and use it in the Authorization header on your API calls.  I don't know if the apikey is tested to have anon or service_role, but by setting the Authorization header the RLS will see that and the apikey is just used to access the project.
https://supabase.com/docs/learn/auth-deep-dive/auth-deep-dive-jwts

**Reply by MariuzM** [reply]
Oh i see, to sum up if i understand correctly. 
- Mint / create jwt token with `authenticated` role inside
- Sign it with the secret jwt key i have in dashboard. 
- Use RLS for auth check to access table

**Reply by GaryAustin1** [reply]
You MAY have to set the Authorization header with the header option on createClient.  I've not tested a custom jwt as the apikey, but using a user jwt does not work for the apikey.    So apikey is still anon, Authorization header is `bearer JWT`.

Not sure what "auth check" is.  If your TO role in the policy is `authenticated` then yes.  If it is auth.uid() = user_id then no as you don't have a user id in the jwt sub field that would match a row's user_id in the table.

**Reply by utgarda** [reply]
I tried a custom JWT for a new role, different from  `service_role`  and `anon`, it does not work as API key at the moment:
```js
{
  message: 'Invalid API key',
  hint: 'Double check your Supabase `anon` or `service_role` API key.'
}
```

**Reply by utgarda** [reply]
`Bearer JWT` auth header works.

```sql
grant my_service_role TO authenticator;
grant USAGE on schema my_private_schema to my_service_role;
grant SELECT, INSERT on all tables in schema my_private_schema to my_service_role;
```

then I use anon API key and auth header with a custom token:

```typescript
    const options: SupabaseClientOptions<'my_private_schema'> = {
        db: { schema: 'my_private_schema' },
        global: {
            headers: {
                Authorization:
                    'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
            },
        },
    }
    supabase = createClient<Database>(SUPABASE_URL, SUPABASE_ANON_KEY, options)
```
